File: app/src/main/java/com/jackyli/androidarchitecture/base/BaseActivity.java
package com.jackyli.androidarchitecture.base;

import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.ViewDataBinding;

public abstract class BaseActivity<T extends ViewDataBinding> extends AppCompatActivity {

    public T binding;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = initLayout();
        initView();
    }

    public abstract void initView();

    public abstract T initLayout();
}

File: app/src/main/java/com/jackyli/androidarchitecture/base/BaseActivity.kt
package com.jackyli.androidarchitecture.base

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.ViewDataBinding

public abstract class BaseActivity<T : ViewDataBinding> : AppCompatActivity() {

    public lateinit var binding: T

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = initLayout()
        initView()
    }

    public abstract fun initLayout(): T

    public abstract fun initView()
}